src/data/dataloader.cpp

```diff
@@ -0,0 +1,693 @@
+
+    template <typename T>
+    using SaveCacheFunc = std::function<void(const T&, std::filesystem::path)>;
+
+    template <typename T>
+    using LoadDataFunc = std::function<T(
+        std::filesystem::path, std::optional<openspace::dataloader::DataMapping> specs
+    )>;
+
+    template <typename T>
+    T internalLoadFileWithCache(std::filesystem::path filePath,
+                                std::optional<openspace::dataloader::DataMapping> specs,
+                                LoadDataFunc<T> loadFunction,
+                                LoadCacheFunc<T> loadCacheFunction,
+                                SaveCacheFunc<T> saveCacheFunction)
+    {
+        static_assert(
+            std::is_same_v<T, openspace::dataloader::Dataset> ||
+            std::is_same_v<T, openspace::dataloader::Labelset> ||
+            std::is_same_v<T, openspace::dataloader::ColorMap>
+        );
+
+        std::string info;
+        if (specs.has_value()) {
+            info = openspace::dataloader::generateHashString(*specs);
+        }
+        std::filesystem::path cached = FileSys.cacheManager()->cachedFilename(filePath, info);
+
+        if (std::filesystem::exists(cached)) {
+            LINFOC(
+                "DataLoader",
+                fmt::format("Cached file {} used for file {}", cached, filePath)
+            );
+
+            std::optional<T> dataset = loadCacheFunction(cached);
+            if (dataset.has_value()) {
+                // We could load the cache file and we are now done with this
+                return *dataset;
+            }
+            else {
+                FileSys.cacheManager()->removeCacheFile(cached);
+            }
+        }
+        LINFOC("DataLoader", fmt::format("Loading file {}", filePath));
+        T dataset = loadFunction(filePath, specs);
+
+        if (!dataset.entries.empty()) {
+            LINFOC("DataLoader", "Saving cache");
+            saveCacheFunction(dataset, cached);
+        }
+
+        return dataset;
+    }
+} // namespace
+
+namespace openspace::dataloader {
+
+namespace data {
+
+Dataset loadFile(std::filesystem::path path, std::optional<DataMapping> specs) {
+    ghoul_assert(std::filesystem::exists(path), "File must exist");
+
+    std::ifstream file(path);
+    if (!file.good()) {
+        throw ghoul::RuntimeError(fmt::format("Failed to open data file {}", path));
+    }
+
+    std::string extension = path.extension().string();
+    for (size_t i = 0; i < extension.size(); i++) {
+        extension[i] = static_cast<char>(tolower(extension[i]));
+    }
+
+    Dataset res;
+    if (extension == ".csv") {
+        res = csv::loadCsvFile(path.string(), specs);
+    }
+    else if (extension == ".speck") {
+        res = speck::loadSpeckFile(path.string(), specs);
+    }
+    else {
+        LERRORC("DataLoader", fmt::format(
+            "Could not read data file {}. File format {} is not supported",
+            path, path.extension()
+        ));
```

Comment:
Yea, we also use it for if-like switch cases. Like:

```cpp
if (string == "abc") {
  return 1;
}
else if (string == "def") {
  return 2;
}

// shouldn't get here
throw ghoul::MissingCaseException();
```

Reply:
👍

Reply:
Although, do we really want to add try-catch blocks everywhere we use any of the `loadCachedFile` or `loadFile` functions...?   That would be a lot of code and added complexity for something that is already handled in the code, and would lead to a fatal crash as soon as someone forgets to check it

Reply:
For the `loadCachedFile`, _that_ funciton should have a try/catch inside and if there is an exception being thrown it should signal back that the cached file wasn't avaiable (i think it's a boolean return value?).  For the `loadFile` I think an exception would be great since it gets caught higher up the chain and logged.

And for the example above, if there is a user-facing way to make `string` not `abc` or `def`, then it should _not_ be a MissingCaseException.  MissingCaseException derives from `std::logic_error` which means that something in the _logic_ of the program went wrong, not something that can be a `std::runtime_error`

Reply:
Conversely, there should never be a `catch` for a `MissingCaseException` since we want those to fail catastrophically to fix the underlying logic error

Reply:
So, the conclusion here is that since the comment here is on the check for the file type, which is the user-facing input, we should not throw a `MissingCaseException`.   Correct?

Also, there are actually cases where we allow calling the `loadFile` function directly (when we don't want to use the caching). So we would still have to catch in manually in some cases

Reply:
But rather do a throw when the loading of a cached file fails :) (currently it returns an `std::nullopt` in those scenarios)

Reply:
I think returning an `std::nullopt` is better since the caching-failing is not necessarily an exceptional behavior (for example if the version has changed, it's the expected behavior).

So I'd say neither to both exceptions

Reply:
Good. I agree.  was a bit confused by this conversation 😅 

I'll consider this resolved by leaving it as is then.